Make ImmutableValidator more precise while analyzing branches

StrongerXi opened the pull request:
## Motivation

Addresses #12864

## Mechanism

1. Delay error generation. Originally we issue error as soon as we see any write to immutable state variable inside a branch, that's can be too conservative, e.g., when both branches write to the variable.
2. Collect all immutable state variable written in a branch (and their position, obtained via the write expression). After analyzing both branch, we can inspect the set difference, and determine which ones are not written (thus initialized) in both paths. Then we issue error for those variables, _and_ collect the all written to continue this process while avoiding duplicated error generation.

## Backward Compatibility

This shouldn't break backward compatibility because we are generating less errors, i.e., correct programs remain correct. The error message is changed, but we can discuss and tweak that.

## Test Case

```solidity
contract Immutable {
    uint256 public immutable iamImmutable;

    constructor(bool shouldInitWithZero) {
        if (shouldInitWithZero) {
            if (shouldInitWithZero) {
                iamImmutable = 0;
            } else {
                iamImmutable = 1;
            }
        } else {
            if (true) {
                iamImmutable = 1;
            }
        }
    }
}
```
This generates a single error at the last write to `iamImmutable`.
If we remove the last conditional `if (true)`, no error is issued.

## Alternative

I actually think the Java way of handling such errors is worth to consider, i.e., it's no less precise than what this PR aims to achieve, and it issues a single error message at the variable declaration site, saying `the final field ... may not have been initialized`. Which is similar to error 2658 in Solidity.

This approach makes the algorithm a bit simpler -- we just need to keep track and filter out the variables-not-initialized-in-both-branches when we issue error 2658, but we lose the location of the problematic write.

StrongerXi committed:
Make ImmutableValidator more precise while analyzing branches

chriseth commented:
Thank you for the pull request! I fear that the complete solution is a bit more complicated. Can you just ignore all the cases after the `else if (m_inBranch)`?

StrongerXi commented:
> I fear that the complete solution is a bit more complicated

Would you mind elaborating on what edge case I might be missing? Based on my limited understanding of Solidity, it seems sound if we take the alternative approach, i.e., mark any variable written only in 1 of the if/else branches to be "not properly initialized".

> Can you just ignore all the cases after the else if (m_inBranch)?

I might be misunderstanding, do you mean deleting all these chunks?
```c++
		else if (m_initializedStateVariables.count(&_variableReference))
		{
			if (!read)
				m_errorReporter.typeError(
					1574_error,
					_expression.location(),
					"Immutable state variable already initialized."
				);
			else
				m_errorReporter.typeError(
					2718_error,
					_expression.location(),
					"Immutable variables cannot be modified after initialization."
				);
		}
		else if (read)
			m_errorReporter.typeError(
				3969_error,
				_expression.location(),
				"Immutable variables must be initialized using an assignment."
			);
		m_initializedStateVariables.emplace(&_variableReference)
```

StrongerXi commented:
Rebase.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,49 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 {
-	bool prevInBranch = m_inBranch;
+	auto prevInBranch = m_inBranch;
+	auto prevWrites = std::move(m_immutableStateVariablesWrittenInBranch);
```
It seems dangerous to me to use `std::move()` here. You also access the moved variable a a few lines down and then move it again, yet to my understanding variables that have been moved should no longer be used...

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,49 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 {
-	bool prevInBranch = m_inBranch;
+	auto prevInBranch = m_inBranch;
+	auto prevWrites = std::move(m_immutableStateVariablesWrittenInBranch);
```
It seems dangerous to me to use std::move() here. You also access the moved variable a a few lines down and then move it again, yet to my understanding variables that have been moved should no longer be used...

StrongerXi replied:
I thought about that at first, then [this post](https://stackoverflow.com/questions/7027523/what-can-i-do-with-a-moved-from-object) convinced me that it's safe to use `clear` on it, which would put it back into a valid state.

The motivation behind using a `std::move` here is to avoid unnecessary copying. It'll be a no-issue if we use an immutable map here.

Regardless, I'm still waiting to hear feedback on the alternative approach, since it'll lead to simpler code and better error message IMO.

Marenz replied:
True, with the info in your link, the usage of `std::move` here seems okay.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,49 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 bool ImmutableValidator::visit(IfStatement const& _ifStatement)
 {
-	bool prevInBranch = m_inBranch;
+	auto prevInBranch = m_inBranch;
```
Why the change away from `bool`?

StrongerXi replied:
Good point. Reverted it to keep changes minimal

chriseth commented:
The main issue I have with this PR is that I'm not convinced that it properly detects multiple writes to a variable in the same branch. Reporting the error about "not initialized" on the variable seems fine to me.

chriseth commented:
The main issue I have with this PR is that I'm not convinced that it properly detects multiple writes to a variable in the same branch. Reporting the error about "not initialized" on the variable seems fine to me.

StrongerXi commented:
@chriseth @Marenz I updated using the alternative approach. The extra check in `ImmutableValidator::checkAllVariablesInitialized` should make the high-level logic pretty clear.

Regarding proper detection of "multiple writes to a variable in the same branch", I believe error 2718 already takes care of it. I also added a test case to illustrate that.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -212,11 +233,7 @@ void ImmutableValidator::analyseVariableReference(VariableDeclaration const& _va
-				_expression.location(),
-				"Cannot write to immutable here: Immutable variables cannot be initialized inside an if statement."
-			);
+			m_immutableStateVariablesWrittenInBranch.emplace(&_variableReference);
```
I think if this is not always an error case, we cannot continue with "else if".

Can you create a test case that detects this bug?

StrongerXi replied:
Thanks for catching this; I think I have a better understanding of your earlier concerns now.

This PR essentially (1) relaxes the multiple write check to allow single write in each branch, and (2) checks for at least a single write in each branch. I made necessary changes for (1) and augmented test case to demonstrate that it falls back to other existing checks in relaxed cases.

StrongerXi commented:
Rebase.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -105,16 +106,40 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 	_ifStatement.trueStatement().accept(*this);
 
+	auto then_writes = std::move(m_immutableStateVariablesWrittenInBranch);
+	decltype(m_immutableStateVariablesWrittenInBranch) else_writes;
```
We're using this set template at least three times. I would suggest to make an alias for it and use it  here and in the class declaration..

Marenz replied:
if we have an alias we can also use it for `prevWrites` instead of `auto`

StrongerXi replied:
Good call, done.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -105,16 +106,40 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 	m_inBranch = prevInBranch;
+	m_immutableStateVariablesWrittenInBranch = std::move(prevWrites);
+	std::vector<ASTNode::SourceLocation const*> bad_write_locs;
+	for (auto var: then_writes) {
```
Please keep using the style with `{` on a new line

StrongerXi replied:
Done.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -105,16 +106,40 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+		{
+			m_immutableStateVariablesWrittenInBranch.emplace(var);
+			m_initializedStateVariables.emplace(var);
+		} else {
```
same here, 
```suggestion
		}
		else
		{
```

StrongerXi replied:
Done.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -280,3 +310,11 @@ void ImmutableValidator::visitCallableIfNew(Declaration const& _declaration)
 }
+
+
+bool ImmutableValidator::mustHaveBeenWrittenInCurrentControlFlow(VariableDeclaration const& _variableReference) const {
```
```suggestion
bool ImmutableValidator::writtenInCurrentControlFlow(VariableDeclaration const& _variableReference) const {
```

This name seems easier to read to me and conveys the same information

StrongerXi replied:
Done.

Marenz commented on test/libsolidity/syntaxTests/immutable/conditionally_initialized.sol:
```diff
@@ -1,9 +1,83 @@
 contract C {
-    uint immutable x;
+    uint immutable singleBranchWrite; // error
```
The test case is comprehensive, but it would be nice to have one test case for each variable-test here, instead of one large one.
Easier to read, faster to see what fails, etc..

StrongerXi replied:
Agree, done.

Marenz reviewed (changes requested):
I didn't find any code logic errors, mostly style.
Oh, a changelog entry is also missing.
Other than that, looks good to me.

StrongerXi commented:
> I didn't find any code logic errors, mostly style. Oh, a changelog entry is also missing. Other than that, looks good to me.

Thanks!

1. All comments are addressed
2. Changelog entry added
3. `ImmutableValidator.h` documentation updated.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -105,16 +106,43 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+	m_immutableStateVariablesWrittenInBranch.clear();
 	_ifStatement.trueStatement().accept(*this);
 
+	auto then_writes = std::move(m_immutableStateVariablesWrittenInBranch);
```
I didn't catch it earlier, but we use `camelCase` here, so `thenWrites` would be the correct style, same for `elseWrites`

StrongerXi replied:
Done.

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -280,3 +313,11 @@ void ImmutableValidator::visitCallableIfNew(Declaration const& _declaration)
 }
+
+
+bool ImmutableValidator::writtenInCurrentControlFlow(VariableDeclaration const& _variableReference) const {
```
Style `{`

StrongerXi replied:
Done

Marenz commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -105,16 +106,43 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+			m_stateVariablesNotInitializedInBothBranches.emplace(var);
+		}
+	}
+	for (auto var: else_writes) {
```
`{`

StrongerXi replied:
Done

StrongerXi commented:
1. Addressed all comments.
2. Added comments to help clarify code logic.
3. Rebase.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,47 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 {
+	// store old context variables
 	bool prevInBranch = m_inBranch;
+	VariableDeclarationSet prevWrites = std::move(m_immutableStateVariablesWrittenInBranch);
```
At this point, `m_immutableStateVariablesWrittenInBranch` is in an undefined state. We should clear it an assert it is cleared after the `accept`. Why is it safe to move it here anyway? Can't there be a write operation inside the if condition?

Marenz replied:
That's not strictly true. It is in an `unspecified` state. 

from https://stackoverflow.com/questions/7027523/what-can-i-do-with-a-moved-from-object:

> Objects of types defined in the C++ standard library may be moved from (12.8). Move operations may be explicitly specified or implicitly generated. Unless otherwise specified, such moved-from objects shall be placed in a valid but unspecified state.

The main question now is if `emplace` is an operation that requires preconditions or not. I am still researching that question

Marenz replied:
Or another phrasing from cppreference.com:

> in a "valid but unspecified state", meaning the object's class invariants hold (so functions without preconditions, such as the assignment operator, can be safely used on the object after it was moved from):

I interpret that, that it is completely okay to use `emplace` or `insert` after it was moved. I am not aware of any `valid` state that would prevent insertions in a `std::set` (e.g. you can't do anything to an instance to prevent `emplace` from working except memory exhaustion)

Marenz replied:
That said, having an explicit clear after the move would def. make it failsafe and will avoid the next code reader wondering the same thing.

StrongerXi replied:
Added explicit `clear` for applicable cases.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,47 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+	// restore and update context
 	m_inBranch = prevInBranch;
+	m_immutableStateVariablesWrittenInBranch = std::move(prevWrites);
+	std::vector<ASTNode::SourceLocation const*> bad_write_locs;
```
```suggestion
	vector<ASTNode::SourceLocation const*> badWriteLocations;
```
(style guide also says "no abbreviations", or at least keep them to a minimum)

chriseth replied:
Oh it's actually not even used?

StrongerXi replied:
My bad -- left from old commit. Removed.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,47 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 	m_inBranch = prevInBranch;
+	m_immutableStateVariablesWrittenInBranch = std::move(prevWrites);
+	std::vector<ASTNode::SourceLocation const*> bad_write_locs;
+	for (auto var: thenWrites)
+	{
+		if (elseWrites.count(var))
+		{
+			m_immutableStateVariablesWrittenInBranch.emplace(var);
+			m_initializedStateVariables.emplace(var);
+		}
+		else
+		{
+			m_stateVariablesNotInitializedInBothBranches.emplace(var);
+		}
+	}
+	for (auto var: elseWrites)
+	{
+		if (!thenWrites.count(var))
+			m_stateVariablesNotInitializedInBothBranches.emplace(var);
+	}
```
```suggestion
	for (auto var: thenWrites)
		if (elseWrites.count(var))
			m_initializedStateVariables.emplace(var);
	m_stateVariablesNotInitializedInBothBranches += (elseWrites - thenWrites) + (thenWrites - elseWrites);
```

StrongerXi replied:
Done -- also added `m_initializedStateVariables.emplace(var);` in the if branch.

chriseth commented on libsolidity/analysis/ImmutableValidator.h:
```diff
@@ -66,17 +66,21 @@ class ImmutableValidator: private ASTConstVisitor
 
-	std::set<VariableDeclaration const*, ASTNode::CompareByID> m_initializedStateVariables;
+	VariableDeclarationSet m_initializedStateVariables;
+	VariableDeclarationSet m_stateVariablesNotInitializedInBothBranches;
```
Wouldn't `m_stateVariablesInitializedInOnlyOneBranch` be a better name?

chriseth replied:
(Variables not initialized in any of the branches are not part of the set)

StrongerXi replied:
Good call, done.

chriseth commented:
Does this need updated documentation?

leonardoalt reviewed (changes requested):
I'm revoking the approval so that this doesn't get accidentally merged since there are still reviewer questions/comments.

Marenz commented:
@StrongerXi would you like to address the open comments or should we take the PR over from here?

StrongerXi commented:
> @StrongerXi would you like to address the open comments or should we take the PR over from here?

My apologies. Got very busy during the last few weeks. Will push a commit by end of today.

Marenz commented:
> My apologies. Got very busy during the last few weeks. Will push a commit by end of today.

All good, just pinging to see whats up :)

StrongerXi commented:
> Does this need updated documentation?

I already updated the documentation in the header.

StrongerXi commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -258,14 +286,22 @@ void ImmutableValidator::checkAllVariablesInitialized(solidity::langutil::Source
+						varDecl->nameLocation(),
+						"Immutable is not initialized in every branch of control flow."
+					);
+				else if (util::contains(m_stateVariablesNotInitializedInBothBranches, varDecl) ||
```
Removed `util::contains(m_stateVariablesNotInitializedInBothBranches, varDecl)` -- it's redundant

Marenz commented:
> I already updated the documentation in the header.

In this case "documentation" means the files under `docs/`, our user-side documentation

StrongerXi commented:
> > I already updated the documentation in the header.
> 
> In this case "documentation" means the files under `docs/`, our user-side documentation

I see. Just did some `grep`s in `docs/` and the only relevant section seems to be [this](https://github.com/ethereum/solidity/blob/develop/docs/contracts/constant-state-variables.rst#immutable). It doesn't mention anything about initialization constraints tho.

What do you think we should do here?

StrongerXi commented:
Rebase.

nikola-matic commented on test/libsolidity/syntaxTests/immutable/conditionally_initialized_multiple_writes.sol:
```diff
@@ -0,0 +1,14 @@
+        } else {
+            multipleWrites = 1;
+        }
+        multipleWrites = 2; // error
```
Is there a "happy case" for this, i.e. a test like this one here, except without the third write, e.g.

```
contract C {
    uint immutable multipleWrites;

    constructor() {
        if (true) {
            multipleWrites = 1;
        } else {
            multipleWrites = 2;
        }
    }
}
```

StrongerXi replied:
Good call. Updated all test cases to add such cases.

StrongerXi commented:
Address comment, rebase, & update changelog.

nikola-matic commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,39 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 	_ifStatement.trueStatement().accept(*this);
 
+	// find out immutable state variables that are written in then/else branches
+	auto thenWrites = std::move(m_immutableStateVariablesWrittenInBranch);
```
@chriseth is not a big fan of generous `auto` usage, and whilst I somewhat disagree with him, it's part of our code style, so please unless the type is present in the statement, e.g. `auto const ptr = std::make_shared<Type>(value)`, replace all instances of `auto` with the actual type.

StrongerXi replied:
Done

nikola-matic commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,39 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 
+	// restore and update context
 	m_inBranch = prevInBranch;
+	m_immutableStateVariablesWrittenInBranch = std::move(prevWrites); // last use of `prevWrites`
```
So I know @Marenz had a concern with the usage of `std::move` up top, but in this case I would prefer to simply do an  assignment
```c++
m_immutableStateVariablesWrittenInBranch = prevWrites;
```
as we might add code later on that will access `prevWrites` after the move, causing us to spend time figuring out why we're getting undefined behaviour.

StrongerXi replied:
This is a meant to be an efficient "restoration" of the `m_immutableStateVariablesWrittenInBranch` context member variable. So `prevWrites` will not (and should not) be used for any other purposes.

nikola-matic commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -104,17 +105,39 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+	for (auto var: thenWrites)
+		if (elseWrites.count(var))
+		{
+			m_immutableStateVariablesWrittenInBranch.emplace(var);
```
Not sure what the type of `var` is (hence the `auto` comment), but unless it's using perfect forwarding, this is gonna be equivalent to `insert(...)`. This is a nit though, so feel free to ignore this comment even if I'm right.

StrongerXi replied:
Agree, changed `auto` to explicit type and `emplace` to `insert`.

nikola-matic commented on test/libsolidity/syntaxTests/immutable/conditionally_initialized_nested_branch_write.sol:
```diff
@@ -0,0 +1,29 @@
+    }
+}
+
+contract C2 {
```
Please move the happy case to a separate file.

StrongerXi replied:
Done.

nikola-matic reviewed (changes requested):
I believe you also need to add a boost test to `test/libsolidity/analysis/FunctionCallGraph.cpp`, but I may be mistaken. See `BOOST_AUTO_TEST_CASE(immutable_initialization)`.

StrongerXi commented:
> I believe you also need to add a boost test to `test/libsolidity/analysis/FunctionCallGraph.cpp`, but I may be mistaken. See `BOOST_AUTO_TEST_CASE(immutable_initialization)`.

Thanks, done.

StrongerXi commented:
Offline chat -- a more general approach is preferred instead of covering control flow one-by-one (e.g., another case would be #13433).

nikola-matic reviewed (commented):
Hey @StrongerXi, so we've discussed this on our team language design call, and the opinions are somewhat split regarding whether to merge this in, or close it in lieu of a more general solution. Most are fine with merging it in, but since @ekpyron would prefer a general approach, and he's off until October, we'll have to wait for his approval, so if we forget, ping us from October 1st to resume discussion.

On the other hand, we're happy with the contribution, so feel free to pick up another issue in the meantime if you feel like it :)

ekpyron commented:
I'm sorry about not doing this sooner, but especially since we will replace the immutable logic by a code data location that will not require the write-once restriction in the constructor at all, the added complexity here is not worth the temporary benefits it may or may not bring.

So thank you very much for the contribution and for your efforts, and again I apologize for not clarifying this sooner, but I'll be closing this PR now.
